### PR TITLE
AADUserInfo Function

### PR DESCRIPTION
**Pull request by dicolanl**
Fixes #
#961  

## Proposed Changes
Function app to pull in AAD User Info daily.

### CODE REVIEW

File: DataConnectors/AADUserInfo/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                        "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING": "[concat('DefaultEndpointsProtocol=https;AccountName=', toLower(parameters('FunctionName')),';AccountKey=', listKeys(resourceId('Microsoft.Storage/storageAccounts', toLower(parameters('FunctionName'))), '2019-06-01').keys[0].value, ';EndpointSuffix=core.windows.net')]",
+                        "WEBSITE_CONTENTSHARE": "[toLower(parameters('FunctionName'))]",
+                        "workspaceID": "[parameters('workspaceID')]",
+                        "workspaceKey": "[concat('@Microsoft.KeyVault(SecretUri=', reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('FunctionName'), 'workspaceKey')).SecretUriWithVersion, ')')]",
+                        "WEBSITE_RUN_FROM_PACKAGE": "https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/AADUserInfo/AADUserInfo_template.zip?raw=true"                    }
```
Comment (v-ampami) [inline DataConnectors/AADUserInfo/azuredeploy.json:197]: Please replace WEBSITE_RUN_FROM_PACKAGE value with [https://aka.ms/sentinel-aaduserinfo-functionapp](https://aka.ms/sentinel-aaduserinfo-functionapp)

File: Parsers/AADUserInfo/AADUserInfo.txt [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Usage Instruction : 
+// Paste below query in log analytics, click on Save button and select as Function from drop down by specifying function name and alias as AADUserInfo.
+// Function usually takes 10-15 minutes to activate. You can then use function alias from any other queries (e.g. AADUserInfo | take 10).
+// Reference : Using functions in Azure monitor log queries : https://docs.microsoft.com/azure/azure-monitor/log-query/functions
+AADUserInfo_CL
```
Comment (v-ampami) [inline Parsers/AADUserInfo/AADUserInfo.txt:5]: Sample data not provided. Getting error ("'summarize' operator: Failed to resolve table or column expression named 'AADUserInfo_CL'") in parser deployment in log analytics workspace. 

Please fix the issue.

**Review by v-ampami — CHANGES_REQUESTED** [review]

**Review by v-maudan — CHANGES_REQUESTED** [review]

File: DataConnectors/AADUserInfo/AADUserInfo/run.ps1 [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    $response = Invoke-RestMethod -Uri $uri -Method $method -ContentType $contentType -Headers $headers -Body $body
+    #$response = Invoke-WebRequest -Uri $uri -Method $method -ContentType $contentType -Headers $headers -Body $body -UseBasicParsing
+    return $response.StatusCode
+
+}
+
+# Get Managed Service Identity info from Azure Functions Application Settings
+$msiEndpoint = $env:MSI_ENDPOINT
```
Comment (v-ampami) [inline DataConnectors/AADUserInfo/AADUserInfo/run.ps1:66]: $msiEndpoint, $msiSecret variables are not used anywhere in the code or taken from input. Can you please remove these variable if you are not using them.

**Comment by v-ampami** [comment]
@dicolanl Please update the following requested changes:

Please move these files to [Solution folder](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions). Follow this [folder structure](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/Box) for reference.

Please include DataConnector file in PR. Refer to this [https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template](https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template) for reference.

Please update changes related to Azure gov. Refer to this [PRLink](https://github.com/Azure/Azure-Sentinel/pull/1762/files) for gov changes.

File: Parsers/AADUserInfo/AADUserInfo.txt (hunk 2 above)
Comment (dicolanl) [reply Parsers/AADUserInfo/AADUserInfo.txt]: added CL schema

**Comment by dicolanl** [comment]
> @dicolanl Please update the following requested changes:
> 
> Please move these files to [Solution folder](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions). Follow this [folder structure](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/Box) for reference.
> 
> Please include DataConnector file in PR. Refer to this https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template for reference.
> 
> Please update changes related to Azure gov. Refer to this [PRLink](https://github.com/Azure/Azure-Sentinel/pull/1762/files) for gov changes.

moved to solutions folder
added sample data
added _CL schema

**Comment by v-ampami** [comment]
> > @dicolanl Please update the following requested changes:
> > Please move these files to [Solution folder](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions). Follow this [folder structure](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/Box) for reference.
> > Please include DataConnector file in PR. Refer to this https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template for reference.
> > Please update changes related to Azure gov. Refer to this [PRLink](https://github.com/Azure/Azure-Sentinel/pull/1762/files) for gov changes.
> 
> moved to solutions folder
> added sample data
> added _CL schema

Please include DataConnector file in PR. Refer to this https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template for reference.

Please fix the PR Validation issues.

File: Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        -customerId $customerId `
+        -sharedKey $sharedKey `
+        -date $rfc1123date `
+        -contentLength $contentLength `
+        -method $method `
+        -contentType $contentType `
+        -resource $resource
+    
+        # Compatible with previous version
+		if ([string]::IsNullOrEmpty($logAnalyticsUri)){
```
Comment (v-ampami) [inline Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1:53]: Please update the logAnalyticsUri code  referring to this [PRLink](https://github.com/Azure/Azure-Sentinel/pull/1762/files).

**Review by v-ampami — CHANGES_REQUESTED** [review]

File: Solutions/AADUserInfo/Data Connectors/README.MD [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+5. Update the value using the string copied from KeyVault.
+* @Microsoft.KeyVault(SecretUri=https://<dnsname>/secrets/workspaceKey/<versionstring>)
+6. Click Ok.
+7. Click edit next to workspaceId.
+8. Update the value with your Sentinel Workspace Id.
+9. Click Ok.
+11. Click Save.
+
+If successfully deployed you should start to see events appear in your Azure Sentinel workspace as soon as they are generated.
+If you run into issues there are a number of options for [monitoring](https://docs.microsoft.com/en-us/azure/azure-functions/functions-monitoring?tabs=cmd) and [deugging](https://docs.microsoft.com/en-us/azure/azure-functions/functions-debug-powershell-local) your Function App.
```
Comment (v-ampami) [inline Solutions/AADUserInfo/Data Connectors/README.MD:83]: This documentation link has en-us locale that is causing the Validation to fail. Please update the link.

**Review by v-ampami — CHANGES_REQUESTED** [review]

File: .vscode/tasks.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (v-ampami) [inline .vscode/tasks.json:1]: Please remove this file from PR.

File: Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1 [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+# Get the current universal time in the default string format
+$currentUTCtime = (Get-Date).ToUniversalTime()
+
+# The 'IsPastDue' porperty is 'true' when the current function invocation is later than scheduled.
+if ($Timer.IsPastDue) {
+    Write-Host "PowerShell timer is running late!"
+}
+
+Import-Module AzureAD -UseWindowsPowerShell
```
Comment (v-ampami) [inline Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1:12]: Getting error while testing. Please fix the issue.
image

**Review by v-ampami — CHANGES_REQUESTED** [review]

File: Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1 (hunk 7 above)
Comment (dicolanl) [reply Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1:12]: How are you testing this? in Azure Functions?  I had to add -UseWindowsPowershell to get it to work in Azure Functions.  There is even a GH issue (here)[https://github.com/Azure/azure-functions-powershell-worker/issues/232] that says to do this in azure functions

File: Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1 (hunk 4 above)
Comment (dicolanl) [reply Solutions/AADUserInfo/Data Connectors/AADUserInfo/run.ps1:53]: @v-ampami. this code looks right.  Line 76 has the env variable, and lines 78-84 check that its a valid url.  this is like many of the other functions already approved?

**Pull request merged**